exe/WrapperLspMain.hs

```diff
@@ -0,0 +1,246 @@
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module WrapperLspMain where
+
+import           Control.Concurrent.STM                (retry)
+import           Control.Monad.Extra
+import           Control.Monad.IO.Class
+import           Control.Monad.IO.Unlift
+import           Control.Monad.Trans.Reader
+import           Data.Aeson
+import           Data.Default
+import           Data.Maybe
+import qualified Data.Set                              as Set
+import qualified Data.Text                             as T
+import           Development.IDE.Core.IdeConfiguration (parseConfiguration)
+import           Development.IDE.LSP.Server            (ReactorChan,
+                                                        ReactorMessage (ReactorNotification, ReactorRequest))
+import qualified Development.IDE.Main                  as Main
+import           Development.IDE.Types.Logger
+import qualified Development.IDE.Types.Logger          as G
+import qualified Development.IDE.Types.Options         as Ghcide
+import qualified Ide.Arguments                         as IdeArgs
+import           Ide.Logger
+import           Ide.Plugin.Config
+import           Language.LSP.Server
+import qualified Language.LSP.Server                   as LSP
+import           Language.LSP.Types
+import qualified System.Directory                      as IO
+import           System.IO
+import qualified System.Log                            as L
+import           UnliftIO
+import           UnliftIO.Concurrent
+
+type ServerM c = ReaderT ReactorChan (LspM c)
+
+lspMain :: IdeArgs.GhcideArguments -> T.Text -> IO ()
+lspMain ghcideArgs@IdeArgs.GhcideArguments{..} msg = do
+    whenJust argsCwd IO.setCurrentDirectory
+    dir <- IO.getCurrentDirectory
+    LSP.setupLogger argsLogFile ["hls"]
+      $ if argsDebugOn then L.DEBUG else L.INFO
+
+    when (Main.isLSP argsCommand) $ do
+        hPutStrLn stderr "Starting (haskell-language-server-wrapper)LSP server..."
+        hPutStrLn stderr $ "  with arguments: " <> show ghcideArgs
+        hPutStrLn stderr $ "  in directory: " <> dir
+
+    let realArguments = def
+          { Main.argCommand = argsCommand
+          , Main.argsLogger = pure hlsWrapperLogger
+          , Main.argsThreads = if argsThreads == 0 then Nothing else Just $ fromIntegral argsThreads
+          , Main.argsIdeOptions = \_config sessionLoader ->
+            let defOptions = Ghcide.defaultIdeOptions sessionLoader
+            in defOptions
+                { Ghcide.optShakeProfiling = argsShakeProfiling
+                , Ghcide.optTesting = Ghcide.IdeTesting argsTesting
+                }
+          }
+
+
+    inH <- Main.argsHandleIn realArguments
+    outH <- Main.argsHandleOut realArguments
+
+    runLanguageServer (Main.argsLspOptions realArguments) inH outH (Main.argsDefaultHlsConfig realArguments) getConfigFromNotification $ \env exitFun -> do
```

Comment:
Duplicating `runLanguageServer` here is not great. This code should either:
1.  call the real `runLanguageServer` (you can always parameterise it even more) 
2. fork it (as done here) and clean up. 

In other words, I'd prefer if you could clean this up or just call the original

Reply:
calling the original is a bit tricky, as the original `runLanguageServer` handles the initialisation request and sends a response which we probably don't want to... If we don't care about that (when I think about it, we probably don't have to), we can re-use `runLanguageServer` with basically a startup continuation. 

I think we can re-use the code, we would open a shake database, etc... but that doesn't matter too much, right?

Reply:
The shake database thing is just a legacy concept - we are not using Shake anymore so all it is now is a bunch of concurrent mutable hash maps.